Incorporate suite host selection logic

sadielbartholomew opened the pull request:
Close #2323 & partially address #1885 & #988. To cover the **first three major points** in the host-selection master plan outlined by @matthewrmshin, copied verbatim into the following checklist:

1. Using `rose.host_select` as reference, develop new logic in cylc to select
   the best suite host:
   - [x] New `cylc` sub-command to get host metrics in a data structure, e.g.:
     - load (uptime)
     - memory (free)
     - df (disk usage)
   - [x] Logic to interpret returned data structure from each host.
   - [x] Logic to discard bad hosts and rank remaining hosts.
2. New settings in `global.rc`:
   - [x] Hosts (+ ranking method + thresholds) and port range for launching suites.
   - [x] Hosts and port range to scan for running suites.
   - [x] ~~(May~~ replace existing `[suite host scanning]`, `[communication]base port`
     and `[communication]maximum number of ports` settings.~~)~~
3. - [x] Modify start up logic of `cylc run` and `cylc restart` to make use of the
   new settings and host select logic where appropriate.
4. - [Later PR] Scheduler main loop to reload `global.rc` at regular intervals. Check if
   current suite host is still a good host for running suites. Shutdown and
   restart itself if required.
5. - [Later PR] Document new behaviour in CUG.
6. - [(n/a)] Remove suite host select logic from Rose.

Whilst this essentially a migration of Rose functionality, the host selection logic was rewritten from ``rose host-select`` in order to be as lightweight / pared down as possible, e.g. by utilising existing cylc methods for remote host management & ssh.

**First round of feedback addressed. Ready for re-review.**

sadielbartholomew commented:
The Codacy failure is now due solely to the Popen module security issue which is also flagged throughout the code base, & in the context of this PR I believe is safe. So (unless anyone thinks otherwise) we can ignore it. Awaiting Travis CI.

I should admit I did a slightly botched rebase including merge conflict resolution :cry:, so if viewing on a commit-by-commit basis note that there may be rogue code in certain commits. In particular the section of code removed from ``remote.py`` in ``5c4f359`` was originally (pre-rebase) removed in ``8a81858``.

sadielbartholomew commented:
@matthewrmshin, @hjoliver: all steps complete, so **open for review**! See my previous comment for explanation RE current test outcomes, which still apply (& note my local test battery also passes).

sadielbartholomew commented:
(New merge conflicts resolved)

hjoliver commented:
@sadielbartholomew - sorry for the review delay, crazy busy here - will attempt next week if this is good to go?

sadielbartholomew commented:
No problem at all @hjoliver I know you generally are super busy! Yes this is good to go.

hjoliver commented:
I've started looking at this.

The PR description needs modifying to not close automatically 988, as a critical part of that is to be done in another PR (item 4 - auto shutdown and restart).

hjoliver commented on bin/cylc-get-host-metric:
```diff
@@ -0,0 +1,189 @@
+
+   Get metrics for localhost, in the form of a JSON structure with
+   top-level keys as requested via the OPTIONS:
+
```
Should the command name be `get-host-metrics` (plural)?  ... indulging in a bit of pendantry, sorry :grimacing:

hjoliver commented on bin/cylc-get-host-metric:
```diff
@@ -0,0 +1,189 @@
+
+   Note: designed as a light-weight command for localhost; to run on a remote
+         host, run after accessing the remote via existing functionality."""
+
```
(I don't really understand what the "Note:..." means - can you explain?)

sadielbartholomew replied:
Sorry if this isn't clear. I wanted to make it explicit that the command in itself was designed (with capability) only to run on localhost, & if someone wanted to run on a remote host, they would have to e.g. run via ``remote_cylc_cmd()`` or ssh to that host before running the command. In other words that a ``--host`` option or similar was not supported (does that make sense now?)

Re-reading this seems self-evident given the enumerated options which can be taken to be exhaustive. I will remove it.

hjoliver commented on lib/cylc/cfgspec/globalcfg.py:
```diff
@@ -270,17 +271,42 @@
+            'thresholds': vdr(vtype='string', default=None),
+        },
+    },
 }
```
uh-oh, we'd better get this PR merged before #2609 :smiling_imp:

hjoliver commented on lib/cylc/cfgspec/globalcfg.py:
```diff
@@ -371,13 +400,30 @@ def upg(cfg, descr):
-        ['hosts', '__MANY__', 'tail command template']
-    )
+        ['hosts', '__MANY__', 'tail command template'])
+    # Assuming version for merge of associated PR is 7.7.0; TODO verify this.
```
I guess this will be 7.8.0 now.

sadielbartholomew replied:
Good spot, I'll get it edited shortly.

hjoliver commented:
(conflicts due to my "external trigger branch" should be easily solved - keep the changes from both branches)

sadielbartholomew commented:
If you do not mind I will resolve any conflicts after most/all reviews are in (& more conflicts may have emerged) as I find the process pretty painful.

hjoliver commented:
@sadielbartholomew - if I were you I would do the conflict resolution now. I just checked again, and it looks to me like you just need to accept the changes from both branches (i.e. do nothing but delete all the `>>>> ... ==== ... <<<<` markers).  If you wait, it may get much worse than that :grimacing:

sadielbartholomew commented:
Thanks @hjoliver, I saw that this specific conflict resolution is as simple as you said in terms of the combining of the conflicting changes, but it is the general process of conflict resolution over a large number of commits spanning significant dev time that I keep struggling with.

However I discussed this briefly with Matt & he said he will guide me through his own git workflow for this scenario. He seconded resolving conflicts as they emerge instead of leaving it all until the end as best practice. So I will get it done now.

matthewrmshin commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+       configuration specifications."""
+
+    def __str__(self):
+        return "No hosts currently compatible with the global configuration."
```
I wonder if it is worth printing out what's configured.

sadielbartholomew replied:
It would be certainly useful & easy enough to implement, so yes.

matthewrmshin commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,186 @@
+    except Exception as exc:
+        if cylc.flags.debug:
+            raise
+        sys.exit(str(exc))
```
Or the caller can hide the error from this command.

sadielbartholomew replied:
I was just using the standard template for an executable, but should have looked through in more detail to consider whether elements such as this were applicable. Agree RE error exposure, & will remove.

matthewrmshin commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                                         measure)
+            except ValueError:
+                print "Threshold value '%s' of wrong type." % cutoff
+                raise
```
Add error string to `ValueError` or print to STDERR.

sadielbartholomew replied:
Sorry, my error handling for this PR is often not correct, as spotted. Thanks for pointing out these instances.

matthewrmshin commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+            except KeyError:
+                print "ERROR: invalid list of keys %s for structure:\n%s." % (
+                    key, nested_data)
+                raise
```
Add error string to `KeyError` or print to STDERR.

matthewrmshin commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                ret_code = process.returncode
+                if ret_code:
+                    # Can't access data => designate as 'bad' host & ignore.
+                    print "Can't obtain metric data from host '%s'." % host
```
Error should go to STDERR.

sadielbartholomew commented:
(merge conflicts resolved) Thanks @hjoliver, I think I have now got the hang of it just about!

Moving onto addressing the feedback which I am prioritising so should start & complete tomorrow.

sadielbartholomew commented on lib/cylc/scheduler_cli.py:
```diff
@@ -162,6 +162,17 @@ def parse_commandline(is_restart):
+    parser.add_option(
+        "--no-host-select",
+        help="Disable host selection functionality.",
+        action="store_true", default=False, dest="no_host_select")
```
Good point! This is definitely an improvement.

sadielbartholomew commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,186 @@
+    return_code = process.returncode
+    if return_code:
+        raise CalledProcessError(return_code, command_str)
+        process.terminate()
```
True! Good spot.

sadielbartholomew commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,186 @@
+        first_match = re.search(data_pattern, raw_string)
+        return first_match.group(1)
+    except (re.error, AttributeError, IndexError):
+        print "No group(1) for 're.search' using '%s'" % data_pattern
```
There's no reason it *should* fail, but I have read in various places that I should use exceptions 'for exceptional (i.e. unexpected) circumstances'.

sadielbartholomew commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,186 @@
+                  r',\s', r'(\d+\.?\d+)', r'\d+\.?\d+']
+MEMORY_TEMPLATES = ['free -k', r'Mem:\s+(.+)\n', r'\d+\s+\d+\s+(\d+)\s+']
+DISK_TEMPLATES = ['df -Pk', r'Mounted on\n\S+\s+(.+)\n',
+                  r'\d+\s+\d+\s+(\d+)\s+']
```
Very good idea. I was assuming anyone reading the code would be far more familiar & experienced with regular expressions than I, so that they read almost trivially, but it certainly will not hurt to add some comments.

sadielbartholomew commented on bin/cylc-get-host-metric:
```diff
@@ -0,0 +1,189 @@
+
+LOAD_TEMPLATES = ['uptime', (1, 5, 15), r'load average:\s(.+)\n',
+                  r',\s', r'(\d+\.?\d+)', r'\d+\.?\d+']
+MEMORY_TEMPLATES = ['free -k', r'Mem:\s+(.+)\n', r'\d+\s+\d+\s+(\d+)\s+']
```
Double-checked with Matt & this is in fact the row we should use. Well spotted!

sadielbartholomew commented on bin/cylc-get-host-metric:
```diff
@@ -0,0 +1,189 @@
+
+
+LOAD_TEMPLATES = ['uptime', (1, 5, 15), r'load average:\s(.+)\n',
+                  r',\s', r'(\d+\.?\d+)', r'\d+\.?\d+']
```
I did combine the expressions similarly later in the relevant method, but in a hacky way to extract each piece of data individually with 're.search'. In general a comprehensive extraction with 're.match' seems a much better solution. Thanks for the outline.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+
+    CMD_BASE = "get-host-metric"  # 'cylc' prepended by remote_cylc_cmd.
+    try:
+        HOSTS = glbl_cfg().get(['suite servers', 'run hosts'])
```
Agreed, that would be more appropriate.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+            try:
+                measure, cutoff = threshold.split(' ')
+            except (AttributeError, ValueError):
+                print "Invalid threshold component '%s'." % threshold
```
Thanks. I've just rectified this in my commit responding to Matt's feedback (which I tackled first as it is more trivial).

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+        if self.selection_complete(host_list):
+            return self.selection_complete(host_list)
+        elif self.RANK_METHOD == 'random':
+            shuffle(host_list)
```
Thanks for pointing this out - I do not want to be bitten again by the mutable default arguments 'gotcha'. I'll have a think about the best way to approach this.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                    cmd = " ".join((cmd, self.opt_mapping[measure][0]))
+        return cmd
+
+    def remove_bad_hosts(self, cmd_with_opts, mock_stats=False):
```
It is indeed a good idea (I've seen this practice before but never really used it myself). I'll apply it to applicable methods.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                if ret_code:
+                    # Can't access data => designate as 'bad' host & ignore.
+                    print "Can't obtain metric data from host '%s'." % host
+                    raise CalledProcessError(ret_code, cmd_with_opts.split())
```
(discussed offline)

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+    def rank_good_hosts(self, all_host_stats):
+        """Take a dictionary of hosts and corresponding metric data and rank
+           hosts via the specified rank method, returning the most suitable."""
+        hosts_with_values_to_rank = dict((
```
I'll check - it certainly seems plausible & would be clearer.

sadielbartholomew replied:
On closer inspection, the ``rank_good_hosts`` code (at this review stage) was in an utterly incomprehensible format & generally hideous, so apologies for that! I have played around with the expressions there & can't seem to get something similar to your snippet to work. Notably, I need to retain the ranked hosts, not the ranked metrics, so it can't be quite as simple as this.

I have, however managed to completely untangle ``sort_asc_hosts`` which was the main offender in terms of comprehensibility & in combination with some piecewise comments I have added, I think this is now sufficiently readable (see latest commits). But feel free to disagree!

NB: I did find a few fancy solutions using ``operator`` module functions as ``keys`` for ``sorted()`` but they seemed a bit far-fetched.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+    def __init__(self):
+        self.use_disk_path = "/"
+        self.opt_mapping = {
+            "load:1": ("--load", ["load", "1 min"]),
```
I believe I had originally went with the current 'nested' structure based on that suggested (in hindsight as an example) in an initial briefing from Matt, but should have considered the best structure to use before starting to code, not least to save myself some difficultly. I (& Tim 'The Zen of Python' Peters) agree in general, & in context this is simpler & will make the code cleaner. Great idea!

sadielbartholomew commented:
@matthewrmshin & @oliver-sanders I have now responded to your first set of feedback. Note that Codacy is failing for the same reasons as before, as outlined in my comment on June 8th, plus one new warning relating to security with the use of ``random.choice``. There is no security issue in this case so this can be ignored.

Please re-review at your convenience.

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+     Filesystem            1024-blocks      Used Available Capacity Mounted on
+     /dev/md0                    96978     33904   *57856*      37% /boot."""
+LOAD_TEMPLATES = ['uptime', (1, 5, 15), re.compile(
+                  r'load average:\s(\d+\.?\d+),\s(\d+\.?\d+),\s(\d+\.?\d+)')]
```
Looking back at the regex, I think you are trying to make the decimal point optional `\d+\.?\d+`, as `\.?` is followed by `\d+` it requires a second digit to match. Probably not an issue (would uptime ever display integers?) but easily fixed e.g. `[\d\.]+`

sadielbartholomew replied:
Well spotted, I'll look at the docs and see if the decimal place is always included.

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+
+       Return stdout if command successful, else raise CalledProcessError."""
+    # Inputs to process are internal only, so 'shell=True' is safe here.
+    process = Popen(command_str, shell=True, stdin=open(os.devnull),
```
Do we actually need to use `shell=True`?

sadielbartholomew replied:
When I remove it, only the load data via the ``uptime`` command seems to come through (i.e. only the ``--load`` tests pass). So I assume we do? I'll have to have a more detailed look at quite why this is.

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+    sys.exit()
+
+
+def run_command_string(command_str):
```
I think this method does the same thing as `subprocess.check_output` https://docs.python.org/2/library/subprocess.html#subprocess.check_output.

sadielbartholomew replied:
Oh dear, trying to re-inventing the wheel! Very well observed.

sadielbartholomew replied:
Alas, ``check_call`` was not added until v2.7, so we can't use it. There is a ``check_call`` available in 2.6 but it doesn't wait for derivative processes, so doesn't work for these cases. I've have to leave this in.

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+    free_memory_str, = extract_pattern(patt, run_command_string(cmd))
+    try:
+        return {"memory": int(free_memory_str)}
+    except TypeError:
```
Ditto.

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+    av_raw_tuple = extract_pattern(patt, run_command_string(cmd))
+    try:
+        av_floats_tuple = [float(av_raw) for av_raw in av_raw_tuple]
+    except TypeError:
```
If we cant cast the memory as a float can we safely continue?

sadielbartholomew replied:
Indeed, no. Though if there was an error converting to float, this would ultimately indicate an issue with the regular expression handling. I'll have another think about how to handle errors relating to the data extraction.

sadielbartholomew replied:
I've changed the cases as highlighted to simply 'raise' the associated exception, which it turns out would be a ``ValueError`` not a ``TypeError`` (oops!)

oliver-sanders commented on bin/cylc-get-host-metrics:
```diff
@@ -0,0 +1,214 @@
+        if free_space_str:
+            try:
+                path_dict["disk-space:" + path] = int(free_space_str)
+            except TypeError:
```
Ditto.

oliver-sanders commented on lib/cylc/scheduler_cli.py:
```diff
@@ -162,6 +162,17 @@ def parse_commandline(is_restart):
+    parser.add_option(
+        "--host-select",
+        help="Enable host selection functionality.",
+        action="store_true", default=False, dest="host_select")
```
@cylc/core How do we want to go about this?

* I guess host selection should be True by default.
* We should be able to overwrite this with `--host` as required. This would make `--no-host-select` similar to `--host=localhost`.
* Should we exit 1 if both are specified.

sadielbartholomew replied:
(discussed in person)

oliver-sanders replied:
**Suggestion: Always use host selection unless overridden by the `--host` setting.**

The default list of suite servers is `["localhost"]` so this is no change from current behaviour if the default is unchanged.

```bash
# no servers configured in user/site config - use localhost
cylc run <suite>

# servers configured - use host selection
cylc run <suite>

# override host selection by manually specifying a host
cylc run --host=localhost
```

We would no longer require the `--host-select` and `--no-host-select` settings.

This should be fine unless there is any reason why we might want to invoke host selection on the host specified by `--host`? I guess such requirements would probably be mitigated by future architecture anyway?

oliver-sanders replied:
The approach I've taken with the health check is to introduce a "condemned hosts" configuration so that hosts have to be explicitly stated before suites jump ship. I think this removes the requirement to disable the health check. Is there a case where we would want to switch off the server health check?

``--no-host-select`` inferring no health check is somewhat implicit, if required we might want to consider something like ``--no-health-check``?

sadielbartholomew replied:
Initially I implemented the options according to Matt's suggestions, but if this is up for debate, here are my thoughts. It seems to me, in agreement with @oliver-sanders, that the cases are as follows:

#### Launching a suite
1. run on specified ``--host=<HOST>`` (including ``HOST=localhost``)
2. with no specification as in [2]: look at the global config:
* with default config, run on localhost (hence it remains the default host choice overall)
* with multiple hosts specified, use host selection to choose one from those specified. (**Q:** if host selection is disabled & ``--host`` is not provided, how would we choose a host from non-default multi-host config? The only option I see is that localhost is therefore implied.)

#### Host health check / reconsideration*
*probably over-simplified, please correct me if necessary.
1. Look at the health of the hosts in order to reconsider & possibly change run host.
2. Not do any health-checking, remaining on the same host as launched on.

Therefore, could we not limit the options to ``--host``, for the usual specification, enabling future health checking, and ``--fixed-host`` (or similar), launching on a permanent host i.e. with health checking disabled? Otherwise host selection would be enabled (& necessary) by default.

This is essentially that is as suggested by Oliver, but with the ``--no-host-select``/``--no-health-check`` option given what I think is a more intuitive name.

sadielbartholomew replied:
(discussed & decided in person)

hjoliver commented:
(I have read the discussion above, but haven't fully wrapped my head around it yet! I'll take a closer look tomorrow...)

oliver-sanders commented:
We've just had a chat about this here and are happy with:

* Always use automatic host selection (default configuration will run suites on "localhost")
* Unless a host is manually specified using `--host`

It comes down to the way in which #988 is implemented. The way I've approached it is to explicitly specify hosts which are "condemned" in a separate configuration rather than to monitor the list of "allowed" server hosts to see if the one we are running on has been removed. This approach means that we don't need to disable the auto-restart mechanism to run on hosts not specified in the "allowed" servers list (e.g. localhost).

tldr; We don't need any `--host-select`, `--no-host-select`, `--no-health-check` options.

oliver-sanders commented on lib/cylc/remote.py:
```diff
@@ -283,3 +296,544 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+            host_stats = dict(mock_stats)  # Prevent mutable object issues.
+        else:
+            for host in self.HOSTS:
+                if host == 'localhost':
```
I think this could potentially cause problems if one were to specify their localhost by name (rather than as `localhost`).

Here is a diff from #988:

```diff
diff --git a/lib/cylc/hostuserutil.py b/lib/cylc/hostuserutil.py
index b6ec6b27e..966f335ce 100644
--- a/lib/cylc/hostuserutil.py
+++ b/lib/cylc/hostuserutil.py
@@ -148,6 +148,8 @@ class HostUtil(object):
 
     def get_fqdn_by_host(self, target):
         """Return the fully qualified domain name of the target host."""
+        if target == 'localhost':
+            return socket.getfqdn()
         return self._get_host_info(target)[0]
 
     def get_user(self):
```
Which you can use to resolve the value of `localhost` like so:
```python
if get_fqdn_by_host(host) == get_fqdn_by_host('localhost'):
```

sadielbartholomew replied:
Very good point & suggestion. Will amend shortly.

sadielbartholomew replied:
It seems 'localhost' gets resolved through the *existing* ``HostUtil`` methods, without need for the included diff (see latest commit, & note the new unit sub-test checking ``get_fqdn_by_host('localhost')`` fails when checking for the unresolved equivalent ``.get('localhost', False)``.

oliver-sanders commented on lib/cylc/remote.py:
```diff
@@ -193,6 +200,12 @@ def __init__(self, argv=None):
+        # Manage run host; host selection always applied unless '--host' set.
+        if not self.host_specified:
+            self.host = HostAppointer().appoint_host()
+            if self.host == 'localhost':
```
See previous comment.

sadielbartholomew commented:
Up-to-date again with feedback responses. Codacy failure irrelevant, as before.

oliver-sanders commented:
Looks like host selection is running twice, once to pick the host then once when attempting to run on that host.

For example configuring `[suite servers]run hosts` to use two remote servers I would expect three ssh'es:

1. `ssh server1 cylc get-host-metrics`
2. `ssh server2 cylc get-host-metrics`
3. `ssh serverx cylc run`

What I am actually seeing is:

1. `ssh server1 cylc get-host-metrics`
2. `ssh server2 cylc get-host-metrics`
3. `ssh serverx cylc run`
4. `ssh server1 cylc get-host-metrics`
5. `ssh server2 cylc get-host-metrics`

Options to deal with this:

1. `cylc run --host=serverx` (I think this is the preferred option)
2. `ssh serverx cylc run --no-host-select`

oliver-sanders commented:
Discussed in person. Ideal process call:

* `ssh server1 cylc get-host-metrics`
* `ssh server2 cylc get-host-metrics`
* `ssh serverx cylc run`

matthewrmshin commented on lib/cylc/remote.py:
```diff
@@ -62,91 +67,164 @@ def watch_and_kill(proc):
-        elif res:
-            sys.stderr.write('ERROR: remote command failed %d\n' % res)
-        return res
+        stdout = PIPE  # Otherwise None by default.
```
Need to set `stdout = None` otherwise - or the variable will not be defined for use in the next statement.

sadielbartholomew replied:
Sorry Matt, I pushed my recent commits by mistake - they need debugging are definitely not ready for review!

kinow commented on lib/cylc/remote.py:
```diff
@@ -24,16 +24,21 @@
 from pipes import quote
 from posix import WIFSIGNALED
+from random import shuffle, choice
+import unittest
```
Hi @sadielbartholomew ! I'm lurking here, receiving notifications for updates - not that I understand everything here - but also planning to test this pull request later, as I was playing with SSH and cylc anyway.

While quickly going through the changes today, just to have an idea of what the change will look like, I noticed the `unittest` module in this file. Wouldn't it be possible to move the `unittest` import and related classes/calls, to under the `tests/` folder?

By the way, I am reading your comments/docstrings, and they are super clear and useful! Thanks for that!

hjoliver replied:
@kinow - shouldn't you be asleep :grin: I must say I just assumed unittests should be in the source file being tested, but maybe you know better, e.g.: https://stackoverflow.com/a/1341053

oliver-sanders replied:
> shouldn't you be asleep

Both of you!

kinow replied:
@hjoliver was going to, but then had one of these [epiphanies](https://i.redd.it/mdjsol1eiad11.jpg) and one issue takes to another pull request, that takes to another repository... Oh, if other files include both code + tests, then all good for me. I'm used with the tests being located separately from the rest of the code, but if the current tests work, coverage is generated OK, etc, so I wouldn't like to distract @sadielbartholomew from the main work in this issue (apologies if I already did so!) :-)

sadielbartholomew commented:
Thanks for your patience, all. I believe I have now fixed the issues outlined by @oliver-sanders in his comment from August 1st (sorry for a slight mess from excessive commit pushing on the way). **Please continue reviewing/testing!** As that comment demonstrated, there is the potential for subtleties in the host selection & remote-command-running process which I could have missed in self-review.

sadielbartholomew commented:
(Rebased to current master branch & commits squashed to a manageable number of related units.)

matthewrmshin commented:
(Tests not happy. Will take a look together tomorrow.)

oliver-sanders commented on lib/cylc/remote.py:
```diff
@@ -214,72 +304,589 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+        for threshold in raw_thresholds_spec.split(';'):
+            threshold = threshold.strip()
+            try:
+                measure, cutoff = threshold.split(' ')
```
I would be tempted to do:

```python
measure, cutoff = threshold.split()
```

To provide resistance against someone putting multiple spaces in a threshold string.

sadielbartholomew replied:
Good thinking.

oliver-sanders commented on lib/cylc/remote.py:
```diff
@@ -214,72 +304,589 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                elif measure == "memory" or measure.startswith("disk-space"):
+                    cutoff = int(cutoff)
+                else:
+                    raise AttributeError("Invalid threshold measure '%s'." %
```
There are a lot of branches to this method, consider simplifying it. For example does it actually matter if we parse memory as a `float` rather than an `int`? Example:

```python
    thresholds = {}
    for threshold in raw_thresholds_spec.split(';'):
        try:
            measure, cutoff = threshold.strip().split()
            if measure.split(':')[0] not in ['load', 'memory', 'disk-space']
                raise ValueError()
            thresholds[measure] = float(cutoff)
        except ValueError:
            raise ValueError('Invalid suite host threshold "%s"' % threshold)
    return thresholds
```

sadielbartholomew replied:
I agree the branching is rather ugly here. I was trying to provide more detailed exceptions for incorrectly formatted input, but on second inspection a single error message should suffice given the valid format required will otherwise be specified clearly.

sadielbartholomew commented on lib/cylc/remote.py:
```diff
@@ -214,72 +304,589 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                ['suite servers', 'run host select', 'rank'])
+            self.THRESHOLDS = glbl_cfg().get(
+                ['suite servers', 'run host select', 'thresholds'])
+        except KeyError:  # Catch for unittest which updates these internally.
```
Agreed - I always thought it was ugly and far from ideal. During development this was needed, hence I added it in, but I've just checked & it isn't needed now. Good spot!

sadielbartholomew commented on lib/cylc/cfgspec/utils.py:
```diff
@@ -125,3 +125,27 @@ def coerce_type(in_str):
+        if startpoint >= endpoint:
+            raise ValueError("%s >= %s but 'X .. Y' format requires X < Y." %
+                             (startpoint, endpoint))
+    except (AttributeError, TypeError):
```
Correct, neither error is required.

sadielbartholomew commented:
To update everyone, since my last comment I have addressed a number of issues that have transpired:

* fixed the parsing & setting of run ports;
* changed to extracting memory from `/proc/meminfo` directly for consistency instead of using the `free` command which we discovered gave output in different tabular style by distro (or specifically by kernel);
* ensured that in verbose or debug mode (& not otherwise) one sees useful messages for anything of relevance e.g. specification of an invalid host, failure to fetch metric from a host, usual error message plus full traceback for the `EmptyHostList` exception, full commands run (including host run on) & ranking order determined;
* prevented (the potentially intensive, with several run hosts set in the global.rc) `HostAppointer.appoint_host()` from running unnecessarily e.g. when `--host` is specified, even if it is a remote host and `remrun()` is called, & when a suite is already running.

oliver-sanders commented on lib/cylc/scheduler_cli.py:
```diff
@@ -66,18 +68,25 @@ def main(is_restart=False):
+    try:
+        SuiteSrvFilesManager().detect_old_contact_file(args[0])
+    except SuiteServiceFileError as exc:
+        sys.exit(exc)
```
We can remove this logic from `cylc.scheduler` (there is no need to check twice).

sadielbartholomew replied:
Good point, though thinking about it, in rare cases wouldn't it be possible for a suite of the same name to start after this check, say during a host selection over many hosts which I have observed to take an order of magnitude of seconds, or during start-up? In that case it might be safest to keep the supplementary check in Scheduler.

oliver-sanders replied:
Well, yes I guess, though this is true at any point of the start up process, reading in the suite configuration can take in the order of tens of seconds...

oliver-sanders replied:
Poke

matthewrmshin replied:
I think it is OK to leave this for now. It is cheap enough to do this check.

oliver-sanders replied:
I'd rather not leave duplicate code lying about, we won't get around to removing it later.

matthewrmshin commented:
Can't get `tests/cylc-poll/17-pbs-cant-connect.t` to pass. Looks like `cylc jobs-poll` is no longer doing the correct stuff.

oliver-sanders commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+these corespond to all the rank methods accepted under the rank setting
+except for \lstinline=random= which does not make sense as a threshold
+measure). Spaces delimit corresponding measures and their values, while
+semi-colons (optionally with subsequent spaces) delimit each measure-value
```
Out of interest, how does this setting play with multiline strings?

```ini
[suite servers]
    [[run host select]]
        thresholds = """
            load: 15 1
            memory: 200000
        """
        
```

sadielbartholomew replied:
``HostAppointer.parse_thresholds`` relies on splitting threshold-cutoff pairs by semi-colon, so that syntax would not work, but I can tweak the code to support it if you think it would be worthwhile (is this a common syntax for compound global config strings)?

oliver-sanders replied:
The `"""` syntax should work for all strings? I'm guessing the previous example would have worked if the thresholds were semi-colon delimited.

sadielbartholomew replied:
Yes the semi-colons are the crucial part. Multi-line strings would not be a problem.

oliver-sanders commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+pair.
+\item {\em options:}
+    \begin{myitemize}
+    \item {\bf load:1} - load average over 1 minute, in KB, (as given by
```
Load is `waiting processes / number of CPUs`, I think load is typically used as a unitless value.

sadielbartholomew commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+    \begin{myitemize}
+    \item {\bf random} - rank for suitability and select a host at random
+    \item {\bf load:1} - rank for suitability and select the host with the
+lowest load average over 1 minute, in KB, (as given by the
```
True, well spotted (ditto for the equivalent comment). I do know this, yet somehow managed to twice assign units incorrectly...

sadielbartholomew commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+\item {\em default:} \lstinline=random=
+\item {\em options:}
+    \begin{myitemize}
+    \item {\bf random} - rank for suitability and select a host at random
```
I personally think the 'rank' setting subsection should only detail the ranking step, as the run host select setting paragraph details the whole process of exclusion then ranking, but if you think this would be especially clear, I will change it?

oliver-sanders replied:
I assumed that's what "rank for suitability" meant. What was it intended to mean?

sadielbartholomew replied:
That is of course what it means. I'm a little confused, I thought you were suggesting to add in the text you commented for this quote? I am saying I do not think it would not really be appropriate for this section, but if you feel strongly I am happy to add it in regardless.

oliver-sanders replied:
It just wasn't clear to me what "rank for suitability" means, if we could clarify a little.

sadielbartholomew commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+    \item {\bf load:5} - as for \lstinline=load:1= above, but over 5 minutes
+    \item {\bf load:15} - as for \lstinline=load:1= above, but over 15 minutes
+    \item {\bf memory} - rank for suitability and select the host with the
+highest usable memory, in KB, as specified under \lstinline=/proc/meminfo=
```
Yes, good idea, I will add that in.

sadielbartholomew commented on doc/src/cylc-user-guide/siterc.tex:
```diff
@@ -748,6 +753,142 @@ \subsection{[hosts]}
+
+\paragraph[thresholds]{[suite servers] \textrightarrow [[run host select]] \textrightarrow thresholds}
+
+A list of metric measure and associated numeric value specifications
```
Agreed, it is a little convoluted. I was trying to get across that there can be thresholds for different measures, but actually that is self-evident for the plural. I'll rephrase in line with your suggestion, thanks.

oliver-sanders replied:
It's worth bearing in mind non-native English speakers and trying to keep language simple.

matthewrmshin commented:
(And `./tests/remote/01-unittests.t`.)

matthewrmshin commented:
For `./cylc-poll/17-pbs-cant-connect.t`, I get these in the suite log:
```
[Errno 2] No such file or directory: '/home/matt/cylc-run/cylctb-20180830T111227Z/cylc-poll/17-pbs-cant-connect/log/job/--host=localhost/job.status'
```

Looks like `--host=localhost` is being used as the task job name for some reason. The arguments are not processed correctly?

sadielbartholomew commented:
@matthewrmshin unless there is something funky/unique about your environment, the test failures you just detailed should now pass, with the underlying issues resolved as far as I can verify.

FYI, for ``17-pbs-cant-connect.t``, I amended the code as necessary only to find that, both on my this branch & my master, I would get the standard MO security message appearing (very inconveniently, I must say) right in the middle of the line that is to be checked to be in the ``log/suite/log``, i.e: 

```
...
	[jobs-poll err]
	WARNING: This computer is provided for the processing of official
	  information. Unauthorised access may constitute a criminal
	  offence. By continuing to use this facility you agree to abide by
	  the Met Office Your Security Commitments. All activity on this
	  system is liable to monitoring.
	
	Connection refused
2018-09-03T18:31:56+01 INFO - [t1.1] -(current:running)(polled) started at 2018-09-03T17:31:32Z
...
```
Accordingly, in my latest commit I have also tweaked the relevant sub-test to search for the contained phrases individually.

matthewrmshin commented:
@hjoliver @oliver-sanders please do the final review.

sadielbartholomew commented:
(Codacy failures are only those that are irrelevant/generic i.e. for use of ``subprocess.Popen`` & ``random.choice``.)

oliver-sanders commented on lib/cylc/cfgspec/globalcfg.py:
```diff
@@ -275,17 +271,17 @@
 
     'suite servers': {
-        'run hosts': vdr(vtype='string_list', default=["localhost"]),
+        'run hosts': vdr(vtype='string_list'),
```
@matthewrmshin 

The removal of the `["localhost"]` default causes traceback if `[suite servers]run hosts` is not defined.

```
TypeError: object of type 'NoneType' has no len()
```

Why did we decentralise this?

oliver-sanders replied:
```
Traceback (most recent call last):
  File "mess", line 5, in <module>
    print ha.appoint_host()
  File "/home/h06/osanders/cylc/lib/cylc/scheduler_cli.py", line 444, in appoint_host
    self.hosts, ignore_if_thresholds_prov=True)
  File "/home/h06/osanders/cylc/lib/cylc/scheduler_cli.py", line 304, in _trivial_choice
    if self.selection_complete(host_list):
  File "/home/h06/osanders/cylc/lib/cylc/scheduler_cli.py", line 285, in selection_complete
    if len(host_list) == 0:
TypeError: object of type 'NoneType' has no len()
```

matthewrmshin replied:
This looks like it is because a null setting of the list here returns a `None` rather than an empty list. I'll add a local fix-and-test against this branch via a PR @sadielbartholomew.

(This default behaviour for settings expecting lists is a bit inconsistent, and can break on dump-and-load of a configuration file. Long term fix expected by #2609.)

sadielbartholomew replied:
Thanks, @matthewrmshin. Note that there is now a new test (just merged in) which will confirm the behaviour is correct with ``[suite servers]run hosts`` not set (I didn't seem to pick up on unfortunately quite obvious this case when writing the original unit test). It currently fails on Travis CI, as well as for me locally.

matthewrmshin replied:
Should be addressed via sadielbartholomew/cylc#5.

matthewrmshin commented on lib/cylc/scheduler_cli.py:
```diff
@@ -328,36 +325,62 @@ def _process_get_host_metrics_cmd(self):
+            if is_remote_host(host):
+                host_proc_map[host] = remote_cylc_cmd(
+                    cmd, stdin=None, host=host, capture_process=True)
+            elif 'localhost' in host_proc_map:
```
This is actually similar to the original logic by @sadielbartholomew but it makes full use of `is_remote_host` to determine if this is `localhost` or not:
* Branch 1: is a remote host
* Branch 2 and 3: is not a remote host, so must be `localhost`.

So as long as we can trust `is_remote_host`, this is the safest way.

matthewrmshin commented:
Actually just spotted a problem (while reading the logic of `rose suite-run`) that needs to be solved as a follow-on.

`rose suite-run` currently exports some custom environment variables on the *remote suite host* before launching `cylc run` (or `cylc restart`). The custom environment variables include e.g. ROSE_VERSION, ROSE_ORIG_HOST, etc and any environment variables defined by the user under the `[env]` section of the suite's `rose-suite.conf`. We need a mechanism to be able to handle this.

oliver-sanders commented:
:fireworks: :champagne: :tada: :heart:
